## Лабораторная работа "Введение в ML"

В этой лабораторной вы:

- познакомитесь с базовыми библиотеками для работы с табличными данными — `numpy` и `pandas`
- поближе посмотрите на простейшие задачи машинного обучения: классификацию и регрессию
- попробуете несколько метрик и поймёте, почему выбор метрики это важно
- обучите несколько простых моделей
- увидите связь между сложностью модели и переобучением
- убедитесь, что без данных всё тлен

Загрузка самых базовых библиотек:

In [1]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
%matplotlib inline

from sklearn.model_selection import train_test_split

### [NumPy](https://numpy.org/doc/stable/user/index.html)

С 1995 numeric, с 2006 NumPy — «Numerical Python extensions» или просто «NumPy»

Возможности библиотеки NumPy:
* работать с многомерными массивами (таблицами)
* быстро вычислять математические функций на многомерных массивах

Ядро пакета NumPy — объект [ndarray](https://docs.scipy.org/doc/numpy/reference/generated/numpy.ndarray.html)

**Важные отличия** между NumPy arrays и Python sequences:
* NumPy array имеет фиксированную длину, которая определяется в момент его создания (в отличие от Python lists, которые могут расти динамически)
* Элементы в NumPy array должны быть одного типа
* Можно выполнять операции непосредственно над NumPy arrays

**Скорость** NumPy достигается с помощью:
* реализации на C
* векторизации и броадкастинга (broadcasting). Например, произведение массивов совместимых форм.

Теперь давайте разберёмся подробнее и сделаем что-нибудь приятное и полезное в `numpy`!

### Индексация

В NumPy работает привычная индексация Python, ура! Включая использование отрицательных индексов и срезов (slices)

<div class="alert alert-info">
<b>Замечание 1:</b> Индексы и срезы в многомерных массивах не нужно разделять квадратными скобками,
т.е. вместо <b>matrix[i][j]</b> нужно использовать <b>matrix[i, j]</b>. Первое тоже работает, но сначала выдаёт строку i, потом элемент j в ней.
</div>

<div class="alert alert-danger">
<b>Замечание 2:</b> Срезы в NumPy создают view, а не копии, как в случае срезов встроенных последовательностей Python (string, tuple and list).
</div>

In [2]:
ones_matrix = np.ones((5, 5))
ones_submatrix_view = ones_matrix[::2,::2] # creates a view, not copy
ones_matrix[::2,::2] = np.zeros((3, 3))
ones_submatrix_view

array([[0., 0., 0.],
       [0., 0., 0.],
       [0., 0., 0.]])

### Ссылка на Яндекс.Контест

Решения и ответы в задачах, расположенных ниже, загружайте в контест на автоматическую проверку:
https://new.contest.yandex.ru/60376/start


**1.** Реализуйте функцию, принимающую на вход два одномерных массива `first_array` и `second_array` и возвращающую матрицу, в которой первый массив соответствует первому столбцу матрицы, второй — второму.

Вероятно первое, что приходит вам на ум, это конкатенация и транспонирование:

In [3]:
def construct_matrix(first_array, second_array):
    """
    Construct matrix from pair of arrays
    :param first_array: first array
    :param second_array: second array
    :return: constructed matrix
    """
    return np.vstack([first_array, second_array]).T # <- your first right code here

In [4]:
construct_matrix(np.array([1,2]),np.array([3,4]))

array([[1, 3],
       [2, 4]])

(в скобках заметим, что конкатенировать можно vertically, horizontally, depth wise методами vstack, hstack, dstack по трём осям (0, 1 и 2, соотвественно), либо в общем случае `np.concatenate` — поиграйтесь ниже с прекрасным примером четырёхмерной точки, чтобы точно всё для себя понять)

In [5]:
p = np.arange(1).reshape([1, 1, 1, 1])
p

array([[[[0]]]])

In [6]:
print("vstack: ", np.vstack((p, p)).shape)
print("hstack: ", np.hstack((p, p)).shape)
print("dstack: ", np.dstack((p, p)).shape)

vstack:  (2, 1, 1, 1)
hstack:  (1, 2, 1, 1)
dstack:  (1, 1, 2, 1)


In [7]:
np.concatenate((p, p), axis=3).shape

(1, 1, 1, 2)

Но, поскольку операция транспонирования [делает массив non-contiguous](https://numpy.org/doc/stable/user/basics.copies.html#other-operations), мы в этой задаче **запретим** ей пользоваться и порекомедуем воспользоваться, например, методом [reshape](https://numpy.org/doc/stable/reference/generated/numpy.reshape.html).

**2.** Реализуйте функцию, принимающую на вход массив целых неотрицательных чисел `nums` и возвращающую самый частый элемент массива.

In [8]:
def most_frequent(nums):
    """
    Find the most frequent value in an array
    :param nums: array of ints
    :return: the most frequent value
    """
    pass # your code here

### Переходим к работе с данными

Прежде всего, загрузим данные и сделаем из них красивые pandas-таблички. Они взяты из параллели RecSys соревнования https://yandex.ru/cup/ml/. Но мы будем иметь дело не со всеми данными, а только с их частью. Данные у нас будут про заведения общественного питания (больше бюрократический терминологии!)

Файлы с данными можно найти [здесь](https://disk.yandex.ru/d/YWvCNRQMb7QSQA).

Задачей будет **предсказание среднего чека** (average_bill) по некоторым другим свойствам заведения.

In [9]:
base = 'DATAS/'

In [10]:
data = pd.read_csv(base + 'organisations.csv')
features = pd.read_csv(base + 'features.csv')
rubrics = pd.read_csv(base + 'rubrics.csv')

В основном мы будем работать с табличкой `data`; остальное вам может пригодиться, если вы захотите знать, какое содержание стоит за кодами признаков.

## Изучение данных

Посмотрите на данные. В этом вам поможет метод ``head`` pandas-таблички.

In [11]:
print(data.head())

                 org_id city  average_bill    rating   rubrics_id  \
0  15903868628669802651  msk        1500.0  4.270968  30776 30774   
1  16076540698036998306  msk         500.0  4.375000        30771   
2   8129364761615040323  msk         500.0  4.000000        31495   
3  15262729117594253452  msk         500.0  4.538813  30776 30770   
4  13418544315327784420  msk         500.0  4.409091        31495   

                                         features_id  
0  3501685156 3501779478 20422 3502045016 3502045...  
1  1509 1082283206 273469383 10462 11617 35017794...  
2  10462 11177 11617 11629 1416 1018 11704 11867 ...  
3  3501618484 2020795524 11629 11617 1018 11704 2...  
4  11617 10462 11177 1416 11867 3501744275 20282 ...  


Полезно посмотреть внимательнее на то, с какими признаками нам предстоит работать.

* **org_id** вам не понадобится;
* **city** - город, в котором находится заведение (``msk`` или ``spb``);
* **average_bill** - средний чек в заведении - он будет нашим таргетом;
* **rating** - рейтинг заведения;
* **rubrics_id** - тип заведения (или несколько типов). Соответствие кодов каким-то человекочитаемым типам живёт в табличке ``rubrics``
* **features_id** - набор неких фичей заведения. Соответствие кодов каким-то человекочитаемым типам живёт в табличке ``features``

Обратите внимание, что **rubrics_id** и **features_id** - это не списки, а разделённые пробелами строки. Когда вам захочется работать с отдельными фичами из мешка фичей для данного заведения, вам придётся всё-таки превратить их в списки (здесь поможет метод `split` для строк).

Чтобы быстро восстанавливать по рубрикам и фичам их нормальные названия, сделайте словари вида ``код_фичи:название_фичи``

In [12]:

rubric_dict = {
    30519: 'Булочная, пекарня',
    30770: 'Бар, паб',
    30771: 'Быстрое питание',
    30774: 'Кафе',
    30775: 'Пиццерия',
    30776: 'Ресторан',
    30777: 'Столовая',
    31286: 'Спортбар',
    31350: 'Кондитерская',
    31375: 'Суши-бар',
    31401: 'Кальян-бар',
    31495: 'Кофейня',
    3108292683: 'Бар безалкогольных напитков',
    3501514558: 'Фудкорт',
    3501750896: 'Кофе с собой'
}
data["rubrics_list"] = data["rubrics_id"].apply(lambda x: [rubric_dict[int(r)] for r in str(x).split() if int(r) in rubric_dict])
print(data["rubrics_list"])

0               [Ресторан, Кафе]
1              [Быстрое питание]
2                      [Кофейня]
3           [Ресторан, Бар, паб]
4                      [Кофейня]
                  ...           
68334                     [Кафе]
68335                 [Ресторан]
68336                     [Кафе]
68337    [Быстрое питание, Кафе]
68338                 [Ресторан]
Name: rubrics_list, Length: 68339, dtype: object


Посмотрим, какими бывают типы заведений:

In [13]:
rubric_dict

{30519: 'Булочная, пекарня',
 30770: 'Бар, паб',
 30771: 'Быстрое питание',
 30774: 'Кафе',
 30775: 'Пиццерия',
 30776: 'Ресторан',
 30777: 'Столовая',
 31286: 'Спортбар',
 31350: 'Кондитерская',
 31375: 'Суши-бар',
 31401: 'Кальян-бар',
 31495: 'Кофейня',
 3108292683: 'Бар безалкогольных напитков',
 3501514558: 'Фудкорт',
 3501750896: 'Кофе с собой'}

Мы что-то поняли про признаки, которыми нам предстоит пользоваться. Теперь время посмотреть на таргет. Вооружившись функциями ``hist`` и ``scatter`` из библиотеки ``matplotlib``, а также методом ``isna`` для pandas-таблиц разберитесь, какие значения принимают таргеты, есть ли там там выбросы, пропуски или ещё какие-то проблемы.

&nbsp;

<details>
  <summary>Когда будете готовы, кликните сюда, чтобы посмотреть ответ</summary>
    <ol>
      <li>Среди таргетов довольно много пропусков;</li>
      <li>Все таргеты - это числа, кратные 500;</li>
      <li>Есть какие-то адские значения, превышающие 100 000 (видимо, выбросы);</li>
      <li>В целом, число ресторанов с данным средним чеком быстро падает с ростом среднего чека. Для средних чеков, больших 2500, заведений уже совсем мало. Примерно у 2/3 заведений средний чек 500.</li>
    </ol>
</details>

In [14]:

missing_count = data["average_bill"].isna().sum()
print(f"Количество пропусков в average_bill: {missing_count}")

Количество пропусков в average_bill: 35561


**Базовая очистка данных**

Раз есть треш, давайте чистить данные.

С пропусками можно бороться по-разному (даже и с пропусками в таргете), но пока мы сделаем самую простую вещь: дропнем все заведения, для которых мы не знаем средний чек.

Уберите из них все заведения, у которых средний чек неизвестен или превышает 2500. Пока есть опасение, что их слишком мало, чтобы мы смогли обучить на них что-нибудь.

**3. Введите в Контест количество заведений, которое у вас получилось после очистки**.

Дальше мы будем работать с очищенными данными.

In [15]:
clean_data = data.dropna(subset=["average_bill"])


clean_data = clean_data[clean_data["average_bill"] <= 2500]

print(f"Количество заведений после очистки: {len(clean_data)}")

Количество заведений после очистки: 32136


**4. Посчитайте и введите в Контест разность между средними арифметическими average_bill в кафе Москвы и Санкт-Петербурга. Округлите ответ до целого.**

&nbsp;

<details>
  <summary>Небольшая подсказка</summary>
  Примените часто используемый метод groupby.
</details>

In [16]:
city_means = clean_data.groupby("city")["average_bill"].mean()

difference = city_means["msk"] - city_means["spb"]

difference_rounded = round(difference)
print(city_means, difference, difference_rounded)

city
msk    792.887230
spb    676.449662
Name: average_bill, dtype: float64 116.43756751957676 116


Давайте ещё немного поизучаем данные. Ответьте на вопросы:

1. Есть ли разница между средними чеками в Москве и Санкт-Петербурге?
2. Коррелирует ли средний чек с рейтингом?
3. Есть ли разница в среднем чеке между ресторанами и пабами (см. соответствующие типы из ``rubrics``)?

&nbsp;

<details>
  <summary>Когда будете готовы, кликните сюда, чтобы посмотреть ответ</summary>
    <ol>
      <li>В целом, да. Вы могли бы сравнить средние (в Москве больше) или медианы (они равны, потому что уж больно много где средний чек 500). Этого, конечно, мало для того, чтобы сделать вывод. Нужно проверять какие-то статические критерии, которые изучаются в курсе по статистике. Не будем останавливаться на этом подробно. Поскольку данные совсем не нормальные, никакой t-тест не сработает; мы бы предложили использовать критерий Манна-Уитни (см. википедию и функцию mannwhitneyu из библиотеки scipy.stats).</li>
      <li>Какая-то корреляция между ними есть но уж больно неубедительная (рекомендуем построим на одном графике boxplot рейтинга по каждому значению среднего чека для визуализации). Конечно, дна становится меньше с ростом среднего чека, но, видимо, в предсказании это особо не используешь;</li>
      <li>Несомненно, в ресторанах средний чек выше. Это и невооружённым глазом видно, и с помощью критерия Манна-Уитни можно проверить.</li>
    </ol>
</details>

## Формулируем задачу

Прежде, чем решать задачу, её надо сформулировать.

**Вопрос первый**: это классификация или регрессия? Подумайте над этим.

&nbsp;

<details>
  <summary>Когда будете готовы, кликните сюда, чтобы посмотреть ответ</summary>
    Ответ не столь однозначен, как хотелось бы. С одной стороны, таргет принимает всего четыре значения, и потому это может быть классификацией с 4 классами. С другой стороны, таргеты - это не абстрактные "треугольник", "круг", "квадрат", а вещественные числа, и когда мы вместо 500 предсказываем 2500, это явно хуже, чем вместо 1500 предсказать 2000. В целом, задачу можно решать и так, и так; мы будем смотреть на метрики обеих задач.
</details>

**Вопрос второй**: какие метрики мы будем использовать для оценки качества решения? Какие метрики вы предложили бы для этой задачи как для задачи классификации? А для этой задачи, как для задачи регрессии?

&nbsp;

<details>
  <summary>Когда будете готовы, кликните сюда, чтобы посмотреть ответ</summary>
    
    Начнём с классификации. Метрика accuracy не очень хороша из-за несбалансированности классов. Действительно, классификатор, который всегда говорит 500, будет иметь accuracy примерно 0.66, хотя это никак не отражает практическую ценность модели. Как мы увидим, самая большая проблема будет заключаться в том, чтобы научиться выделять заведения с большими чеками, а их меньше всего и в accuracy они вносят самый маленький вклад. Есть разные способы с этим бороться, один -- использовать sklearn.metrics.balanced_accuracy_score. Его идея, грубо говоря, в том, чтобы по каждому классу найти, какая доля объектов этого класса правильно классифицирована, а потом эти доли усреднить. Тогда у бессмысленного классификатора, который всем ставит 500, будет скор 1/5 (ведь классов 5), а чтобы получить прежние 2/3, нужно будет научиться в каждом классе правильно ставить хотя бы 2/3 меток.    
    
    Теперь что касается регрессии. Основых метрики две - MSE и MAE. Из первой стоит извлекать корень, чтобы получать интерпретируемые человеком значения, а вторая менее агрессивна к выбросам (впрочем, выбросов тут уже нет, мы их все выкинули). Без дополнительной информации не очень понятно, какую выбирать, можно брать любую. А выбирать надо: ведь даже банальные модели "предсказывай всегда среднее" и "предсказывай всегда медиану" будут по-разному ранжироваться этими метриками.
    
</details>

**Вопрос третий**: а не взять ли нам какую-нибудь более экзотическую метрику? Например, MAPE (определение в учебнике в главе про оценку качества моделей). А как вам такое соображение: допустим, заказчик говорит, что пользователи будут расстраиваться, только если мы завысили средний чек - так давайте поправим MSE или MAE, обнуляя те слагаемые, для которых предсказанный таргет меньше истинного. Вот это хорошая метрика или нет?

&nbsp;

<details>
  <summary>Когда будете готовы, кликните сюда, чтобы посмотреть ответ</summary>
    
    Что касается MAPE, у нас нет тех проблем, с которой она борется. Вот если бы у нас были средние чеки от 500 до миллиона, мы бы столкнулись с ситуацией, что большие ошибки для больших чеков доминировали бы в сумме для MSE и MAE (500 вместо 1000 меркнет по сравнению с 500к вместо миллиона). Говоря поэтически, мы бы оптимизировали модель для миллионеров, забыв про простых трудяг. И было бы логично перейти от парадигмы "ошибаемся на 500 рублей" к парадигме "ошибаемся на 50%". Но у нас все таргеты примерно одного порядка, MAPE нам особо ни к чему.
    
    Вторая метрика коварна тем, что её можно "накрутить" безо всякой пользы для дела. А именно, модель, которая всегда предсказывает средний чек в миллион, была бы идеальна. Но все бы расстраивались и не ходили есть. Другое дело, что можно ввести разные веса для ошибок в большую и в меньшую сторону, но опять же - пока нет показаний к тому, что это нужно.
    
</details>

## Применяем ML

Теперь время разбить данные на обучающую и тестовую выборку. Делается это с помощью функции ``train_test_split`` из пакета ``sklearn``. При этом очень важно сделать две вещи:

* Зафиксировать ``random_state=42`` (да, именно этот, а то ваши модели могут не зайти в Контест), чтобы всё, что мы делаем, было воспроизводимо (иначе от перезапуска к перезапуску числа могут меняться, и мы не будем понимать, из-за чего это происходит).
* Сделать стратификацию по таргету. В противном случае у нас в трейне и тесте могут оказаться разные пропорции классов (обычно особенно страдают мало представленные классы), что неутешительно скажется на результате.

**Обратите внимание**, что если вы побьёте выборку на train и test по-другому, ваши результаты могут не зайти в контест.

In [17]:
clean_data_train, clean_data_test = train_test_split(
    clean_data, 
    stratify=clean_data['average_bill'], 
    test_size=0.33, 
    random_state=42
)

print("Train size:", len(clean_data_train), "Test size:", len(clean_data_test))


Train size: 21531 Test size: 10605


Теперь нам нужен **бейзлайн** - очень простая модель, с которой мы в дальнейшем будем сравниваться.

Поскольку мы ещё не знаем никаких умных классов моделей, все модели мы будем писать руками. А именно, мы напишем две простых модели на основе ``sklearn.baseRegressorMixin`` и ``sklearn.base.ClassifierMixin`` (посмотрите примеры в документации sklearn и сделайте так же):

* Модель для задачи регрессии, которая для всех заведений предсказывает одно число — среднее значение среднего чека;
* Модель для задачи классификации, которая для всех заведений предсказывает один класс — самый частый класс (ироничным образом он в данном случае совпадает с медианой).

**Важно!** Мы будем много раз повторять вам мантру о том, что **информация из тестовой выборки не должна протекать в процесс обучения**. Так вот, и среднее, и самый частый класс вы должны считать именно на обучающей выборке!

**5 и 6. Напишите эти две модели и сдайте в Контест**. В процессе проверки модели будут и обучаться, и предсказывать.

Заметим, что для этих моделей нам вообще не нужны какие-то "фичи"; мы работаем только с таргетом.

У каждой модели есть (как минимум) два метода: `fit` (обучает модель по фичам `X` и таргету `y`) `predict` (предсказывает по фичам `X`)

In [18]:
from scipy.stats import mode

from sklearn.base import RegressorMixin

class MeanRegressor(RegressorMixin):
    # Predicts the mean of y_train
    def fit(self, X=None, y=None):
        '''
        Parameters
        ----------
        X : array like, shape = (n_samples, n_features)
        Training data features
        y : array like, shape = (_samples,)
        Training data targets
        '''
        self.mean_ = np.mean(y)
        return self  # возвращаем self, чтобы в дальнейшем можно было писать reg = MeanRegressor().fit(...)

    def predict(self, X=None):
        '''
        Parameters
        ----------
        X : array like, shape = (n_samples, n_features)
        Data to predict
        '''
        """
        Предсказываем сохранённое среднее для любой выборки
        """
        n_objects = X.shape[0] if X is not None else 1
        return np.full(shape=(n_objects,), fill_value=self.mean_)

from sklearn.base import ClassifierMixin

class MostFrequentClassifier(ClassifierMixin):
    # Predicts the rounded (just in case) median of y_train
    def fit(self, X=None, y=None):
        '''
        Parameters
        ----------
        X : array like, shape = (n_samples, n_features)
        Training data features
        y : array like, shape = (_samples,)
        Training data targets
        '''
        """
        Запоминаем самый частотный класс из y на обучающей выборке
        """
        # способ через np.unique:
        values, counts = np.unique(y, return_counts=True)
        self.most_freq_class_ = values[np.argmax(counts)]
        return self

    def predict(self, X=None):
        '''
        Parameters
        ----------
        X : array like, shape = (n_samples, n_features)
        Data to predict
        '''
        """
        Предсказываем сохранённый самый частотный класс для любой выборки
        """
        n_objects = X.shape[0] if X is not None else 1
        return np.full(shape=(n_objects,), fill_value=self.most_freq_class_)

Обучим наши модели

In [19]:
reg = MeanRegressor()
reg.fit(y=clean_data_train['average_bill'])

clf = MostFrequentClassifier()
clf.fit(clean_data_train, clean_data_train['average_bill'])

Посмотрим, насколько хорошо справился регрессор:

In [20]:
pred_vals_clf = clf.predict(clean_data_test)

true_vals_clf = clean_data_test['average_bill'].values

for c in np.unique(true_vals_clf):
    mask_class = (true_vals_clf == c)
    total_in_class = mask_class.sum()
    correct_in_class = ((true_vals_clf == pred_vals_clf) & mask_class).sum()
    print(f"Class {c}: total={total_in_class}, correct={correct_in_class}")
    


Class 500.0: total=7368, correct=7368
Class 1000.0: total=1809, correct=0
Class 1500.0: total=890, correct=0
Class 2000.0: total=391, correct=0
Class 2500.0: total=147, correct=0


Обучите модели и оцените их качество на тестовой выборке. В качестве метрик возьмём RMSE (``np.sqrt`` от ``sklearn.metrics.mean_squared_error``) и ``sklearn.metrics.balanced_accuracy_score``.

Для регрессионной модели имеет смысл считать только RMSE (значения будут не кратны 500, точно мы угадывать не будем никогда), а вот для классификационной можно найти обе метрики. Сделайте это. Какая модель оказалась лучше по RMSE?

<details>
  <summary>Когда будете готовы, кликните сюда</summary>
    
  Казалось бы, регрессор никогда не угадывает, но он в каком-то смысле лучше классификатора - справедливо ли это? Возможно. Несуществующий пользователь модели вряд ли будет задавать вопросы "почему средний чек не кратен 500?" Ну, выдали около 800 - ок, понятно.
    
</details>

## Усложнение модели

Бейзлайны будут нашей отправной точкой. Строя дальнейшие модели, мы будем спрашивать себя: получилось ли лучше бейзлайна? Если нет или если не особо, то в чём смысл усложнения?

Начнём с использования фичи ``city``. Мы уже видели, что в разных городах и средние чеки разные. Легко проверить, что *медиана* средних чеков всё же одна и та же и в Москве, и в Санкт-Петербурге (ох уж этот вездесущий средний чек 500!), поэтому с классификатором мы ничего не сделаем. Но вот регрессор можно попробовать починить.

**7. Напишите регрессор, для каждого заведения предсказывающий среднее значение в том же городе (на обучающей выборке, конечно) и сдайте его в Контест**. Вам может помочь то, что булевы `pandas` и `numpy` столбцы можно умножать на численные — в такой ситуации False работает, как ноль, а True как единица.

In [21]:
from sklearn.base import RegressorMixin

class CityMeanRegressor(RegressorMixin):
    def fit(self, X=None, y=None): 
        """
        Запоминаем средний чек для объектов из Москвы и для объектов из Питера.
        """
        # Маски для строк из Москвы и Питера
        mask_msk = (X['city'] == 'msk')
        mask_spb = (X['city'] == 'spb')
        
        self.mean_msk_ = y[mask_msk].mean()  # среднее y по Москве
        self.mean_spb_ = y[mask_spb].mean()  # среднее y по Питеру
        
        
        return self

    def predict(self, X=None):
        """
        Для объектов из Москвы предсказываем self.mean_msk_,
        для объектов из Питера — self.mean_spb_.
        """
        # Пошаговый способ (списки):
        predictions = []
        for city_val in X['city']:
            if city_val == 'msk':
                predictions.append(self.mean_msk_)
            elif city_val == 'spb':
                predictions.append(self.mean_spb_)
        return np.array(predictions)

Обучите регрессор и сравните его по метрике RMSE с бейзлайнами. Получилось ли улучшить метрику?

In [22]:
city_reg = CityMeanRegressor()
city_reg.fit(clean_data_train, clean_data_train['average_bill'])

y_pred_city_reg = city_reg.predict(clean_data_test)
print(y_pred_city_reg)

[677.71044658 792.61119082 792.61119082 ... 792.61119082 792.61119082
 677.71044658]


Лучше стало, но, правда, не очень сильно. В этот момент очень важно не просто радовать руководителя приростом в третьем знаке, но и думать о том, что происходит.

Средний средний чек по Москве равен 793, в Санкт-Петербурге - 676, а в целом - 752 рубля. MSE, увы, не поможет вам ответить на вопрос, стало ли лучше пользователю, если вы ему вместо 752 рублей назвали 793. Здесь вскрывается весьма существенный порок MSE в этой задаче. Дело в том, что наш изначальный таргет делит заведения на некоторые "ценовые категории", и различие в средних чеках 500 и 1000 в самом деле существенно. Наверное, мы хотели бы как раз правильно предсказывать ценовые категории. Но MSE не очень помогает нам об этом судить. Дальше мы ещё подумаем, как это исправить.

В любом случае, несмотря на улучшение метрики, мы пока не можем судить, стало ли по жизни лучше от усложнения модели.

In [23]:
from sklearn.metrics import mean_squared_error
rmse_city_reg = np.sqrt(mean_squared_error(clean_data_test['average_bill'], y_pred_city_reg))
print("CityMeanRegressor RMSE =", rmse_city_reg)


CityMeanRegressor RMSE = 445.1063281403263


Поручинившись немного, возьмём на вооружение другую идею. Давайте использовать типы заведений!

Но с типами есть некоторая проблема: в столбце ``rubrics_id`` не всегда один идентификатор, часто их несколько, и всего комбинаций довольно много. Чтобы не возиться с малочисленными типами, давайте сольём их в один безликий ``other``.

Итак, добавьте в обучающие и тестовые данные столбец ``modified_rubrics``, в котором будет то же, что и в ``rubrics_id``, если соответствующая комбинация рубрик содержит хотя бы 100 заведений из обучающей (!) выборки, и строка ``other`` в противном случае.

Здесь вам поможет контейнер ``Counter`` из библиотеки ``collections``.

In [24]:
from collections import Counter

def build_modified_rubrics(train_df, test_df, threshold=100):
    """
    Создаёт столбец 'modified_rubrics' в обоих датасетах,
    где все нечастые комбинации заменены на 'other'.
    """
    rubr_counter = Counter(train_df['rubrics_id']) #Считаем частоты рубрик в train
    
    def to_modified_rubr(rubr_str):
        count_ = rubr_counter[rubr_str]
        if count_ >= threshold:
            return rubr_str
        else:
            return 'other'
    
    train_df['modified_rubrics'] = train_df['rubrics_id'].apply(to_modified_rubr)
    
    def to_modified_rubr_test(rubr_str):
        if rubr_str not in rubr_counter:  # здесь мы смотрим только на rubr_counter (из train!),
            return 'other'
        elif rubr_counter[rubr_str] >= threshold:
            return rubr_str
        else:
            return 'other'
    
    test_df['modified_rubrics'] = test_df['rubrics_id'].apply(to_modified_rubr_test)
    
    return train_df, test_df


clean_data_train, clean_data_test = build_modified_rubrics(clean_data_train, clean_data_test, threshold=100)


Теперь настало время написать могучий классификатор, который по заведению предсказывает медиану средних чеков среди тех в обучающей выборке, у которых с ним одинаковые `modified_rubrics` и город (вы спросите, почему медиану, а не самый частый -- спишем это на вдохновение; самый частый тоже можно брать - но медиана работает лучше).

**8. Напишите классификатор и сдайте в Контест**.

In [25]:
from sklearn.base import ClassifierMixin
import numpy as np

class RubricCityClassifier(ClassifierMixin):
    def fit(self, X, y):
        """
        Запоминаем, для каждой (city, modified_rubrics) - медиану y,
        а также глобальную медиану на случай отсутствия комбинации.
        """
        self.dict_ = {}
        
        # Склеим X,y в одну таблицу для удобства
        df = X.copy()
        df['average_bill'] = y.values
        
        # Глобальная медиана (по трейну)
        self.global_median_ = np.median(y)
        
        # Группировка по (city, modified_rubrics)
        grouped = df.groupby(['city', 'modified_rubrics'])['average_bill']
        
        # Считаем медиану в каждой группе
        # Это будет pd.Series с MultiIndex [(city, rubrics), ...] и значением медианы
        med_series = grouped.median()
        
        # Превращаем в обычный словарь {(city, rubrics): median_value, ...}
        self.dict_ = med_series.to_dict()
        
        # Подготовим множество возможных дискретных таргетов, чтобы потом округлять
        self.all_bills_ = np.sort(df['average_bill'].unique())
        
        return self
    
    def _round_to_known_bills(self, val):
        """
        Округляет число val до ближайшего из self.all_bills_ (например, [500,1000,1500,...])
        """
        # Поиск индекса в отсортированном массиве self.all_bills_
        idx = np.searchsorted(self.all_bills_, val)
        # idx - это позиция, куда можно вставить val, чтобы сохранить порядок
        # нужно сравнить self.all_bills_[idx-1] и self.all_bills_[idx], где idx может быть 0 или len(array)
        if idx == 0:
            return self.all_bills_[0]
        elif idx == len(self.all_bills_):
            return self.all_bills_[-1]
        else:
            # ближайший из двух соседей
            left_ = self.all_bills_[idx-1]
            right_ = self.all_bills_[idx]
            if abs(left_ - val) <= abs(right_ - val):
                return left_
            else:
                return right_
    
    def predict(self, X):
        """
        Для объекта из (city, modified_rubrics) возвращаем округлённую медиану,
        если связка есть в self.dict_. Иначе берём глобальную медиану.
        """
        preds = []
        for city_val, rubr_val in zip(X['city'], X['modified_rubrics']):
            key = (city_val, rubr_val)
            if key in self.dict_:
                raw_median = self.dict_[key]
            else:
                raw_median = self.global_median_
            # округляем к ближайшему известному таргету
            final_pred = self._round_to_known_bills(raw_median)
            preds.append(final_pred)
        return np.array(preds)

Сравните обученный классификатор по метрикам RMSE и balanced_accuracy_score с нашими бейзлайнами. Получилось ли улучшить?

In [26]:
rubr_city_clf = RubricCityClassifier()
rubr_city_clf.fit(clean_data_train, clean_data_train['average_bill'])

y_pred_rubr_city = rubr_city_clf.predict(clean_data_test)

from sklearn.metrics import balanced_accuracy_score, mean_squared_error

bal_acc = balanced_accuracy_score(clean_data_test['average_bill'], y_pred_rubr_city)
rmse_rc = np.sqrt(mean_squared_error(clean_data_test['average_bill'], y_pred_rubr_city))

print("Rubric+City classifier balanced_accuracy =", bal_acc)
print("Rubric+City classifier RMSE =", rmse_rc)


Rubric+City classifier balanced_accuracy = 0.30552511833185647
Rubric+City classifier RMSE = 393.96675836287915


Обратите внимание что рост accuracy по сравнению с бейзлайном при этом на порядок меньше:

accuracy_score

Predict most frequent:  0.6947666195190948

Predict by rubric and city:  0.7095709570957096

Часто обнаружится, что более редкие классы модель угадывает не очень хорошо.

In [27]:
true_vals = clean_data_test['average_bill'].values
pred_vals = y_pred_rubr_city

for c in np.unique(true_vals):
    mask_class = (true_vals == c)
    total_in_class = mask_class.sum()
    correct_in_class = ((true_vals == pred_vals) & mask_class).sum()
    print(f"Class {c}: total={total_in_class}, correct={correct_in_class}")


Class 500.0: total=7368, correct=6608
Class 1000.0: total=1809, correct=700
Class 1500.0: total=890, correct=217
Class 2000.0: total=391, correct=0
Class 2500.0: total=147, correct=0


Для диагностики напечатайте для каждого класса тестовой выборки, сколько в нём объектов и скольким из них наш классификатор приписал правильный класс. Что вы видите?

&nbsp;

<details>
  <summary>Когда будете готовы, кликните сюда, чтобы посмотреть ответ</summary>
    
  Вы, вероятно, видите то, что мы стали однозначно лучше по сравнению с бейзлайном детектировать средний чек 1000 и 1500 (хотя всё равно не очень хорошо + ценой ухудшения качества на среднем чеке 500), а вот чеки 2000 и 2500 нам ну никак не даются.
    
</details>

**Кстати**. А вы понимаете, почему приведённый выше пайплайн классификации был не очень удачным с точки зрения архитектуры? Почему его было бы правильнее воплотить по-другому?

&nbsp;

<details>
  <summary>Когда будете готовы, кликните сюда, чтобы посмотреть ответ</summary>
Собственно говоря, и не было никакого пайплайна. К счастью, у нас была одна обучающая выборка, мы на ней посчитали список рубрик для modified_rubrics и радовались жизни. Но если бы нам надо было переобучать всё на новых данных, пришлось бы помнить, что их надо везде пересчитать (ведь у нас могли появиться новые рубрики с хотя бы 100 представителями). А уж никакую кросс-валидацию (кто знает - тот поймёт) с нашим подходом к делу и вовсе бы не получилось сделать без боли.
    
Поэтому в следующей лабораторной вы научитесь делать честные пайплайны, в которых преобразование данных, генерация фичей и обучение классификатора будут объединены в один понятный процесс, происходящий на этапе fit.
</details>

## Слишком простые и слишком сложные модели

Бейзлайны у нас слишком просты и потому не очень полезны в жизни. Но если сложность модели растёт бесконтрольно, то тоже получается плохо.

Давайте рассмотрим конкретный пример. Создадим классификатор, использующий одновременно `rubrics_id` и `features_id`.

Сделайте следующее:

- для каждого объекта обучающей выборки сконкатенируйте строку `rubrics_id` с разделителем (например, буквой 'q') и содержимым `features_id`. Полученный столбец озаглавьте `modified_features`. Это не самый клёвый способ заиспользовать все фичи, но сейчас пока сойдёт. Причём на сей раз не будем выкидывать мало представленные значения (вся информация важна, не так ли?).
- при этом для тестовой выборке заменяйте на строку `other` все конкатенации, которые не встретились в обучающей выборке.

То есть элементы в этом столбце будут иметь вид `other` или `30776 30774 q 3502045032 11741 3502045016 1046...`.

Формируем "modified_features" в train

Обучаемся

Теперь обучите классификатор, который для заведения предсказывает медиану среднего чека по всем объектам тестовой выборки с таким же, как у него, значением `modified_features`, а если такого в обучающей выборке нет, то глобальную медиану среднего чека по всей обучающей выборке.

**9. Загрузите в Контест предсказания этого классификатора на тестовой выборке**

Мы ждём файла **.csv**, у которого в каждой строке будет только одно число - предсказание классификатора.

Возможно, вам будет полезна библиотека ``tqdm``, позволяющая отслеживать в реальном времени, сколько времени уже крутится цикл и сколько итераций ещё осталось. Впрочем, если вы всё написали нормально, то должно работать не очень долго.

Предсказываем на тесте и выгружаем в .csv:

In [28]:

def make_rubfeat_str(rubrics_id, features_id):
    return rubrics_id + 'q' + features_id

train_medians_dict = {}

tmp_map = {}

for i, row in clean_data_train.iterrows():
    combo_str = make_rubfeat_str(row['rubrics_id'], row['features_id'])
    if combo_str not in tmp_map:
        tmp_map[combo_str] = []
    tmp_map[combo_str].append(row['average_bill'])

for key, bills in tmp_map.items():
    train_medians_dict[key] = np.median(bills)

global_median = np.median(clean_data_train['average_bill']) # глобальную медиана, если ничего не найдётся

predictions_task9 = [] # для теста

for i, row in clean_data_test.iterrows():
    combo_str = make_rubfeat_str(row['rubrics_id'], row['features_id'])
    if combo_str in train_medians_dict:
        pred_val = train_medians_dict[combo_str]
    else:
        if "other" in train_medians_dict:
            pred_val = train_medians_dict["other"]
        else:
            pred_val = global_median
    predictions_task9.append((row['org_id'], pred_val))

with open('task9_predictions.csv', 'w', encoding='utf-8') as f:
    for org_id_val, pred_val in predictions_task9:
        f.write(f"{org_id_val},{int(pred_val)}\n")

print("Файл сохранён.")


Файл 'task9_predictions.csv' для Контеста сохранён.


Модель, очевидно, очень сложная. Число параметров (различных категорий) в ней сопоставимо с числом объектов в обучающей выборке. А получилось ли хорошо?

Давайте посчитаем RMSE и balanced_accuracy_score на обучающей и на тестовой выборках.

**10. Введите их в Контест**

Налицо переобучение: на трейне метрики отличные, на тесте - вообще никакие

В общем, не гонитесь за чрезмерной сложностью модели..

Простая функция, которая формирует для df_source предсказания на основе словаря train_medians (rubrics+features -> медиана).
Если нет ключа, пытается 'other', иначе глобальная медиана:

In [29]:
from math import sqrt
def predict_median_rubfeat(df_source, train_medians, global_med):
    preds = []
    for i, row in df_source.iterrows():
        key = make_rubfeat_str(row['rubrics_id'], row['features_id'])
        if key in train_medians:
            preds.append(train_medians[key])
        else:
            if "other" in train_medians:
                preds.append(train_medians["other"])
            else:
                preds.append(global_med)
    return np.array(preds)

Повторим обучение на train:

In [30]:

tmp_map_10 = {}
for i, row in clean_data_train.iterrows():
    combo_str = make_rubfeat_str(row['rubrics_id'], row['features_id'])
    if combo_str not in tmp_map_10:
        tmp_map_10[combo_str] = []
    tmp_map_10[combo_str].append(row['average_bill'])

train_medians_dict_10 = {}
for key, bills in tmp_map_10.items():
    train_medians_dict_10[key] = np.median(bills)

global_median_10 = np.median(clean_data_train['average_bill'])


Предсказываем на train и test:

In [31]:

train_preds_10 = predict_median_rubfeat(clean_data_train, train_medians_dict_10, global_median_10)
test_preds_10  = predict_median_rubfeat(clean_data_test,  train_medians_dict_10, global_median_10)

Считаем RMSE и balanced_acc:

In [32]:
rmse_train_10 = sqrt(mean_squared_error(clean_data_train['average_bill'], train_preds_10))
balacc_train_10 = balanced_accuracy_score(clean_data_train['average_bill'], train_preds_10)

rmse_test_10 = sqrt(mean_squared_error(clean_data_test['average_bill'], test_preds_10))
balacc_test_10 = balanced_accuracy_score(clean_data_test['average_bill'], test_preds_10)

rmse_train_10_rounded = round(rmse_train_10, 2)
balacc_train_10_rounded = round(balacc_train_10, 2)
rmse_test_10_rounded = round(rmse_test_10, 2)
balacc_test_10_rounded = round(balacc_test_10, 2)

print(rmse_train_10_rounded, 
      balacc_train_10_rounded, 
      rmse_test_10_rounded, 
      balacc_test_10_rounded)


32.42 0.99 513.99 0.2


/home/ekaterina/.local/lib/python3.10/site-packages/sklearn/metrics/_classification.py:2524: UserWarning: y_pred contains classes not in y_true
  warnings.warn("y_pred contains classes not in y_true")
/home/ekaterina/.local/lib/python3.10/site-packages/sklearn/metrics/_classification.py:2524: UserWarning: y_pred contains classes not in y_true
  warnings.warn("y_pred contains classes not in y_true")


## ML без данных что компутер без электричества

Возможно, вы смотрите на полученные выше результаты и думаете: вот если бы мы не какие-то убогие медианы предсказывали, а гоняли бы нейросети, то тут-то бы всё и получилось!

Но, увы, совсем даже не всегда от счастья нас отделяет выбор хорошей модели (и стратегии обучения). Если данные не очень, то даже самая крутая модель не сработает. В этой ситуации нужно либо добывать новые фичи каким-то образом, либо собирать новые данные (увеличивать датасет), либо просто бросать задачу.

Давайте посмотрим, что выжмет из наших данных одна из самых мощных моделей для табличных данных - градиентный бустинг на решающих деревьях в исполнении [CatBoost](https://catboost.ai/).

Но прежде, чем сделать fit, нам надо облагородить данные. Несмотря на то, что CatBoost отлично работает с категориальными фичами, мешок признаков из `rubrics_id` или `features_id` может ему оказаться не по зубам. Поэтому мы соберём датасет в пристойную матрицу, создав для каждого типа рубрик и фичей отдельный столбец и записав там единицы для тех объектов, у которых эта рубрика или фича имеет место.

В матрице почти все элементы будут нулями. Такие матрицы считаются **разреженными** и их можно хранить гораздо эффективней, чем просто таблицей. Этим и займёмся)

Есть несколько форматов хранения разреженных матриц (многие из них реализованы в [пакете sparse библиотеки scipy](https://docs.scipy.org/doc/scipy/reference/sparse.html)), и каждый пригоден для чего-то своего.

Создавать разреженную матрицу лучше в [формате COO](https://docs.scipy.org/doc/scipy/reference/generated/scipy.sparse.coo_array.html#scipy.sparse.coo_array). Он предполагает, что разреженная матрица задаётся в виде трёх списков: `row`, `col`, `data`, причём каждая тройка `(row[i], col[i], data[i])` кодирует элемент со значением `data[i]`, стоящий на позиции `(row[i], col[i])`. Считается, что на позициях `(row, col)`, которые ни разу не встретились, стоят нули.

Нетрудно видеть, что заполнять такую матрицу - одно удовольствие, и особенно этому помогает тот факт, что **пара `(row, col)` может встретиться несколько раз** (тогда в итоговой матрице на соответствующей позиции стоит сумма соответствующих `data[i]`). Но, с другой стороны, почти ничего другого с такой матрицей не сделаешь: произвольного доступа к элементам она не предоставляет, умножить её тоже особо ничего не умножишь. Поэтому для дальнейшего использования созданную таким образом матрицу преобразуют в один из более удобных форматов, например, [CSR (compressed sparse row)](https://scipy-lectures.org/advanced/scipy_sparse/csr_matrix.html). Он, к примеру, хорошо подходит для умножения на вектор (потому что матрица хранится по строкам). Не будем разбирать его подробно, но можете почитать по ссылке, если интересно.

Вам нужно будет превратить обучающие и тестовые данные в разреженные матрицы `sparse_data_train` и `sparse_data_test` соответственно, таким образом, что:

- столбец `city` превратится в столбец из единиц и нулей (например, 1 - Москва, 0 - Питер);
- столбец `rating` перекочует в разреженные матрицы без изменений;
- каждый типы рубрик и каждая фича превратятся в отдельный 0-1-принак;

В тестовой выборке будут фичи, которых в обучающей выборке не было. С ними можно по-разному работать, но давайте создадим дополнительную фантомную фичу `feature_other`, в которой будет то, сколько неизвестных по обучающей выборке фичей есть у данного объекта.

In [33]:
import numpy as np
import pandas as pd
from sklearn.model_selection import train_test_split
from collections import Counter
from scipy.stats import mode
from math import sqrt

from sklearn.metrics import mean_squared_error, balanced_accuracy_score

from scipy import sparse


print("Библиотеки успешно установлены и импортированы!")

Библиотеки успешно установлены и импортированы!


In [34]:

def df_to_sparse(df):
    """
    Преобразует DataFrame (с колонками city, rating, rubrics_id, features_id) в CSR-матрицу.
    """
    row_inds = []
    col_inds = []
    data_vals = []
    
    for row_i, row in enumerate(df.itertuples(index=False)):
        
        city_val = 1.0 if row.city == 'msk' else 0.0
        row_inds.append(row_i)
        col_inds.append(0)
        data_vals.append(city_val)
        
        rating_val = row.rating if not np.isnan(row.rating) else 0.0 # rating: если NaN, ставим 0 
        row_inds.append(row_i)
        col_inds.append(1)
        data_vals.append(rating_val)
        
        rubs = row.rubrics_id.split()
        for r in rubs:
            if r in rubrics_index:
                j = rubrics_index[r] + 2  # +2, тк 0 и 1 заняты city/rating
                row_inds.append(row_i)
                col_inds.append(j)
                data_vals.append(1.0)
        
        feats = row.features_id.split()
        for f in feats:
            if f in features_index:
                j = features_index[f] + 2 + len(rubrics_index) # для фич инлекс начинается после всех рубрик
                row_inds.append(row_i)
                col_inds.append(j)
                data_vals.append(1.0)
            else: 
                j = 2 + len(rubrics_index) + feature_other_idx # для фич, которых не было в трейне
                row_inds.append(row_i)
                col_inds.append(j)
                data_vals.append(1.0)
    
    n_rows = len(df)
    n_cols = 2 + len(rubrics_index) + len(features_index) + 1  # +1 для feature_other
    coo_mat = sparse.coo_matrix((data_vals, (row_inds, col_inds)), shape=(n_rows, n_cols))
    
    csr_mat = coo_mat.tocsr()
    return csr_mat

Собираем sparse-матрицы:

In [35]:

sparse_data_train = df_to_sparse(clean_data_train)
sparse_data_test  = df_to_sparse(clean_data_test)

NameError: name 'rubrics_index' is not defined

Данные готовы, и теперь можно запустить катбуст

In [63]:
from catboost import CatBoostClassifier

print("CatBoost успешно установлен и импортирован!")


CatBoost успешно установлен и импортирован!


In [64]:
clf_cat = CatBoostClassifier(verbose=False, random_state=42)
clf_cat.fit(sparse_data_train, clean_data_train['average_bill'])

**11. Пришлите в Контест balanced_accuracy_score на тестовой выборке, округлённый до двух знаков после запятой**. Стало ли сильно лучше от того, что мы воспользовались таким крутым классификатором?

Предсказываем:

In [66]:
pred_cat_test = clf_cat.predict(sparse_data_test)
balacc_cat = balanced_accuracy_score(clean_data_test['average_bill'], pred_cat_test)
balacc_cat_rounded = round(balacc_cat, 2)

print("CatBoostClassifier balanced_accuracy_score на тесте:", balacc_cat_rounded)


CatBoostClassifier balanced_accuracy_score на тесте: 0.36
